File: android/app/src/main/java/com/slywombat/slyled/audio/EnvelopeFollower.kt
package com.slywombat.slyled.audio

import kotlin.math.abs
import kotlin.math.exp
import kotlin.math.sqrt

/**
 * #804 — Pure-DSP envelope follower for Android Auto Brightness.
 *
 * Pipeline per buffer (8 kHz mono, ~50 ms hops):
 *   1. Single biquad low-pass at ~250 Hz isolates kicks/bass.
 *   2. RMS energy of the band-passed slice (the "instantaneous" level).
 *   3. Onset peak-pick: spectral-flux-style positive difference vs. EMA of past RMS.
 *   4. Attack / release one-pole smoother on (RMS + onset bonus).
 *   5. Sensitivity gain → soft clip 0..1 → floor/ceiling map → 0..255.
 *
 * Pure Kotlin; no Android imports → unit-testable on JVM.
 */
class EnvelopeFollower(
    private val sampleRateHz: Int = 8000,
    cutoffHz: Float = 250f,
) {
    // Biquad LPF state (RBJ cookbook coeffs)
    private val b0: Float
    private val b1: Float
    private val b2: Float
    private val a1: Float
    private val a2: Float
    private var z1 = 0f
    private var z2 = 0f

    // Onset detector state — slow EMA of past RMS for spectral-flux baseline.
    private var rmsBaseline = 0f
    private val baselineAlpha = 0.05f

    // Output smoother — separate attack & release time constants.
    private var smoothed = 0f

    var sensitivity: Float = 1.5f
    var floor: Float = 0.05f
    var ceiling: Float = 1.0f
    var attackMs: Float = 8f
    var releaseMs: Float = 220f

    init {
        // RBJ low-pass biquad
        val w0 = 2f * Math.PI.toFloat() * cutoffHz / sampleRateHz
        val cos = kotlin.math.cos(w0)
        val sin = kotlin.math.sin(w0)
        val q = 0.707f
        val alpha = sin / (2f * q)
        val a0 = 1f + alpha
        b0 = ((1f - cos) / 2f) / a0
        b1 = (1f - cos) / a0
        b2 = ((1f - cos) / 2f) / a0
        a1 = (-2f * cos) / a0
        a2 = (1f - alpha) / a0
    }

    /**
     * Push one buffer of mono PCM samples in the range [-1f, 1f] and return
     * the smoothed envelope value 0..1 after this hop.
     */
    fun process(samples: FloatArray, length: Int = samples.size): Float {
        if (length <= 0) return smoothed
        var sumSq = 0f
        for (i in 0 until length) {
            // Biquad direct-form-II transposed
            val x = samples[i]
            val y = b0 * x + z1
            z1 = b1 * x - a1 * y + z2
            z2 = b2 * x - a2 * y
            sumSq += y * y
        }
        val rms = sqrt(sumSq / length)

        // Onset bonus = positive deviation above slow baseline
        val onset = (rms - rmsBaseline).coerceAtLeast(0f) * 2f
        rmsBaseline += baselineAlpha * (rms - rmsBaseline)

        val target = (rms + onset) * sensitivity
        val clipped = softClip(target)

        // One-pole asymmetric smoother. dt is buffer length / sampleRate.
        val dtMs = (length.toFloat() / sampleRateHz) * 1000f
        val tau = if (clipped > smoothed) attackMs else releaseMs
        val a = if (tau <= 0f) 1f else 1f - exp(-dtMs / tau)
        smoothed += a * (clipped - smoothed)
        if (smoothed < 0f) smoothed = 0f
        if (smoothed > 1f) smoothed = 1f
        return smoothed
    }

    /** Map current envelope to a 0..255 master brightness. */
    fun toMaster(env: Float = smoothed): Int {
        val f = floor.coerceIn(0f, 1f)
        val c = ceiling.coerceIn(f, 1f)
        val mapped = f + (c - f) * env.coerceIn(0f, 1f)
        return (mapped * 255f + 0.5f).toInt().coerceIn(0, 255)
    }

    fun reset() {
        z1 = 0f; z2 = 0f
        rmsBaseline = 0f
        smoothed = 0f
    }

    /** Last smoothed envelope value (0..1) without re-processing. */
    val envelope: Float get() = smoothed

    private fun softClip(x: Float): Float {
        // 1 - exp(-x) — smooth knee at 1.0, never exceeds 1.
        if (x <= 0f) return 0f
        return 1f - exp(-abs(x))
    }
}

File: ios/SlyLED/Audio/EnvelopeFollower.swift
// EnvelopeFollower — pure-DSP port of Android's EnvelopeFollower.kt
// (constants must stay byte-for-byte to keep envelope traces matching
// the Kotlin implementation against the shared WAV test corpus).
//
// Pipeline per 50 ms hop (~400 samples @ 8 kHz):
//   1. Single biquad LPF @ 250 Hz isolates kick/bass.
//   2. RMS energy of the band-passed slice.
//   3. Spectral-flux-style onset = (RMS − slow EMA baseline), clamped ≥0,
//      multiplied by 2.0.
//   4. Asymmetric one-pole smoother (attack 8 ms, release 220 ms).
//   5. Sensitivity gain → 1−exp(−x) soft clip → 0..1.

import Foundation

final class EnvelopeFollower {
    // Biquad LPF (RBJ cookbook coeffs)
    private let b0, b1, b2, a1, a2: Float
    private var z1: Float = 0
    private var z2: Float = 0

    private var rmsBaseline: Float = 0
    private let baselineAlpha: Float = 0.05

    private(set) var smoothed: Float = 0

    var sensitivity: Float = 1.5
    var floor: Float = 0.05
    var ceiling: Float = 1.0
    var attackMs: Float = 8
    var releaseMs: Float = 220

    private let sampleRateHz: Int

    init(sampleRateHz: Int = 8000, cutoffHz: Float = 250) {
        self.sampleRateHz = sampleRateHz
        let w0 = 2 * Float.pi * cutoffHz / Float(sampleRateHz)
        let c = cosf(w0)
        let s = sinf(w0)
        let q: Float = 0.707
        let alpha = s / (2 * q)
        let a0 = 1 + alpha
        b0 = ((1 - c) / 2) / a0
        b1 = (1 - c) / a0
        b2 = ((1 - c) / 2) / a0
        a1 = (-2 * c) / a0
        a2 = (1 - alpha) / a0
    }

    @discardableResult
    func process(_ samples: UnsafePointer<Float>, count: Int) -> Float {
        guard count > 0 else { return smoothed }
        var sumSq: Float = 0
        for i in 0..<count {
            let x = samples[i]
            let y = b0 * x + z1
            z1 = b1 * x - a1 * y + z2
            z2 = b2 * x - a2 * y
            sumSq += y * y
        }
        let rms = sqrtf(sumSq / Float(count))

        let onset = max(0, rms - rmsBaseline) * 2.0
        rmsBaseline += baselineAlpha * (rms - rmsBaseline)

        let target = (rms + onset) * sensitivity
        let clipped = softClip(target)

        let dtMs = (Float(count) / Float(sampleRateHz)) * 1000.0
        let tau = (clipped > smoothed) ? attackMs : releaseMs
        let a: Float = tau <= 0 ? 1.0 : 1.0 - expf(-dtMs / tau)
        smoothed += a * (clipped - smoothed)
        if smoothed < 0 { smoothed = 0 }
        if smoothed > 1 { smoothed = 1 }
        return smoothed
    }

    func toMaster(_ env: Float) -> Int {
        let f = max(0, min(1, floor))
        let c = max(f, min(1, ceiling))
        let mapped = f + (c - f) * max(0, min(1, env))
        return min(255, max(0, Int(mapped * 255 + 0.5)))
    }

    func reset() {
        z1 = 0
        z2 = 0
        rmsBaseline = 0
        smoothed = 0
    }

    private func softClip(_ x: Float) -> Float {
        if x <= 0 { return 0 }
        return 1 - expf(-abs(x))
    }
}
